### PR TITLE
SW-278_condition_handler_for_sdym

### PR DESCRIPTION
**Pull request by wladi0097**

### CODE REVIEW

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php [hunk 1]
```diff
@@ -283,4 +279,10 @@ public function addUserGroup($usergroup)
     {
         $this->parameters['usergrouphash'] = StaticHelper::calculateUsergroupHash($this->shopKey, $usergroup);
     }
+
```
Comment (ihanli) [inline FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: Remove one of these empty lines.

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php [hunk 2]
```diff
@@ -283,4 +279,10 @@ public function addUserGroup($usergroup)
...
+
+    public function addFlag(string $key, $value)
```
Comment (ihanli) [inline FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: Interesting. Shouldn't this fail for PHP 5.6 because the string type hint is only supported in PHP 7.x?

File: FinSearchUnified/Resources/services.xml [hunk 3]
```diff
@@ -16,6 +16,12 @@
             <tag name="shopware.event_subscriber" />
         </service>
 
+        <service id="fin_search_unified.sdym_criteria_request_handler"
```
Comment (ihanli) [inline FinSearchUnified/Resources/services.xml]: Put the attribute `id` on a separate line and move both attributes one level to the left (like the other service definitions).

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+use Shopware\Bundle\SearchBundle\Criteria;
+use Enlight_Controller_Request_RequestHttp as RequestHttp;
+use FinSearchUnified\Bundle\SearchBundle\CriteriaRequestHandler\SdymCriteriaRequestHandler;
+use Shopware\Tests\Functional\Bundle\StoreFrontBundle\Helper;
+use Shopware\Tests\Functional\Bundle\StoreFrontBundle\TestContext;
+
+class SdymCriteriaRequestHandlerTest extends TestCase
+{
+    /** @var RequestHttp */
+    public $request;
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: Use this template for your class members:
```
/**
 * @var RequestHttp
 */
public $request;

/**
 * @var Criteria
 */
public $criteria;
```
Mind the separating empty lines between the member definitions.

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    public $criteria;
+    /** @var TestContext */
+    public $context;
+    /** @var SdymCriteriaRequestHandler */
+    public $handler;
+
+    protected function setUp()
+    {
+        parent::setUp();
+
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: Remove one of these empty lines.

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+        $this->criteria = new Criteria();
+        $this->context = $this->getContext();
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: Use the following code to get a context instead of the `getContext` method:
```
$contextService = Shopware()->Container()->get('shopware_storefront.context_service');
$context = $contextService->getShopContext();
```

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                false,
+                'hey'
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider handleDataProvider
+     */
+    public function testHandle($paramValue, $shouldExist, $paramKey = Constants::SDYM_PARAM_FORCE_QUERY)
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: While I can see why you gave `$paramKey` a default value, fixtures and data sets should always be as explicit as possible. So move the default value to the data provider.

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    {
+        $request = new RequestHttp();
+        $request->setParams([$paramKey => $paramValue]);
+
+        $this->handler->handleRequest($request, $this->criteria, $this->context);
+
+        if ($shouldExist) {
+            $this->assertEquals(Constants::SDYM_PARAM_FORCE_QUERY, $this->criteria->getConditions()[0]->getName());
+            $this->assertEquals(SimpleCondition::class, get_class($this->criteria->getConditions()[0]));
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: Generally spoken, the assertion order should be:

1. Check if the condition exists.
2. Check the conditions value.

In this case you could run `$this->assertTrue($this->criteria->hasCondition('forceOriginalQuery'))` which would cover both previous assertions.

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        } else {
+            $this->assertEmpty($this->criteria->getConditions());
+            $this->assertEquals('array', gettype($this->criteria->getConditions()));
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: Why is this assertion necessary?

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+        $contextService = Shopware()->Container()->get('shopware_storefront.context_service');
+        $this->context = $contextService->getShopContext();
+    }
+
+    /**
+     * @throws Exception
+     */
+    public function testGenerateCondition()
+    {
+        $term = 'ye';
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php]: Rename `$term` to `$name`.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+        $handler = new SimpleConditionHandler();
+        $handler->generateCondition(
+            new SimpleCondition($term),
+            $this->querybuilder,
+            $this->context
+        );
+
+        $params = $this->querybuilder->getParameters();
+
+        $this->assertArrayHasKey('ye', $params, '"query" was not found in querybuilder parameters array');
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php]: You will have to come up with your own assertion message since this one doesn't apply to the tested class.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+        $this->assertSame(true, $params['ye'], 'Expected "query" parameter to contain "searchterm"');
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php]: You will have to come up with your own assertion message since this one doesn't apply to the tested class.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php [hunk 13]
```diff
@@ -134,6 +135,16 @@ public function testCreateQueryWithConditions()
         $this->assertCount(3, $attrib, 'Expected attributes to not contain any other parameters');
     }
 
+    public function testSimpleCondition()
+    {
+        $criteria = new Criteria();
+        $criteria->addCondition(new SimpleCondition('ye'));
+        $query = $this->factory->createQuery($criteria, $this->context);
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php]: Let your code breath. Add an empty line before this one.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php [hunk 14]
```diff
@@ -134,6 +135,16 @@ public function testCreateQueryWithConditions()
...
+        $params = $query->getParameters();
+        $this->assertArrayHasKey('ye', $params);
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/QueryBuilderFactoryTest.php]: This could use an empty line too. Add it before this line.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/SearchQueryBuilderTest.php [hunk 15]
```diff
@@ -164,6 +164,42 @@ public function testAddQueryMethod($searchTerm, $expectedResult)
         $this->assertSame($expectedResult, $parameters['query'], sprintf('Expected query to be "%s"', $expectedResult));
     }
 
+    public function queryBuilderAddFlagDataProvider()
+    {
+        return [
+            'value set' => [
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/SearchQueryBuilderTest.php]: Uppercase the first letter of your data sets names.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/SearchQueryBuilderTest.php [hunk 16]
```diff
@@ -164,6 +164,42 @@ public function testAddQueryMethod($searchTerm, $expectedResult)
...
+                'non empty string',
+                true
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider queryBuilderAddFlagDataProvider
+     */
+    public function testQueryBuilderAddFlag($value, $expected)
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/SearchQueryBuilderTest.php]: Add the missing type hints.

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php (hunk 2 above)
Comment (wladi0097) [reply FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: There are a lot of typehints all over the place, i don't know

File: FinSearchUnified/Resources/services.xml [hunk 17]
```diff
@@ -12,16 +12,23 @@
     <services>
         <service id="fin_search_unified.subscriber.frontend" class="FinSearchUnified\Subscriber\Frontend">
             <argument>%fin_search_unified.plugin_dir%</argument>
-            <argument type="service" id="template" />
-            <tag name="shopware.event_subscriber" />
+            <argument type="service" id="template"/>
```
Comment (ihanli) [inline FinSearchUnified/Resources/services.xml]: Nope. Revert these changes. Don't touch anything that's not related to the story.

File: FinSearchUnified/Resources/services.xml [hunk 18]
```diff
@@ -12,16 +12,23 @@
...
+            <tag name="shopware.event_subscriber"/>
+        </service>
+
+        <service
+            id="fin_search_unified.sdym_criteria_request_handler"
+            class="FinSearchUnified\Bundle\SearchBundle\CriteriaRequestHandler\SdymCriteriaRequestHandler"
+        >
+            <tag name="criteria_request_handler"/>
```
Comment (ihanli) [inline FinSearchUnified/Resources/services.xml]: Separate the double quote and slash by a single white space.

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php [hunk 19]
```diff
@@ -283,4 +279,9 @@ public function addUserGroup($usergroup)
     {
         $this->parameters['usergrouphash'] = StaticHelper::calculateUsergroupHash($this->shopKey, $usergroup);
     }
+
+    public function addFlag($key, $value)
```
Comment (ihanli) [inline FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: Since we now don't have any info about the arguments, add PHPDocs.

File: FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php [hunk 20]
```diff
@@ -0,0 +1,96 @@
...
+            'ParamKey is something different' => [
+                true,
+                false,
+                'hey'
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider handleDataProvider
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundle/CriteriaRequestHandler/SdymCriteriaRequestHandlerTest.php]: Add type hints for the argument list.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/SearchQueryBuilderTest.php [hunk 21]
```diff
@@ -164,6 +164,44 @@ public function testAddQueryMethod($searchTerm, $expectedResult)
...
+            'Value is non empty string' => [
+                'non empty string',
+                true
+            ]
+        ];
+    }
+
+    /**
+     * @dataProvider queryBuilderAddFlagDataProvider
+     * @param mixed $value The Flag value
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/SearchQueryBuilderTest.php]: Add an empty line before this one.

File: FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php [hunk 22]
```diff
@@ -0,0 +1,57 @@
...
+        $handler = new SimpleConditionHandler();
+        $handler->generateCondition(
+            new SimpleCondition($name),
+            $this->querybuilder,
+            $this->context
+        );
+
+        $params = $this->querybuilder->getParameters();
+
+        $this->assertArrayHasKey($name, $params, '"ye" was not found in the params');
```
Comment (ihanli) [inline FinSearchUnified/Tests/Bundle/SearchBundleFindologic/ConditionHandler/SimpleConditionHandlerTest.php]: * `Expected parameter for simple condition not found`
* `Expected parameter to be true`

File: FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php [hunk 23]
```diff
@@ -283,4 +279,14 @@ public function addUserGroup($usergroup)
     {
         $this->parameters['usergrouphash'] = StaticHelper::calculateUsergroupHash($this->shopKey, $usergroup);
     }
+
+    /**
+     * Add a key - value pair to the parameters
```
Comment (ihanli) [inline FinSearchUnified/Bundle/SearchBundleFindologic/QueryBuilder.php]: Doesn't add any value. I'm satisfied with the presence of those `@param`s.

File: FinSearchUnified/Resources/services.xml (hunk 18 above)
Comment (ihanli) [reply FinSearchUnified/Resources/services.xml]: The change described in the previous comment didn't happen.

Comment (wladi0097) [reply FinSearchUnified/Resources/services.xml]: My bad, I thought the revert of this file would fix this.
